src/coin_test/analysis/graphs.py

```diff
@@ -0,0 +1,516 @@
+        fig.update_layout(
+            title={"text": title, "font": plot_params.title_font},
+            xaxis_title={"text": x_lbl, "font": plot_params.axes_font},
+            yaxis_title={"text": y_lbl, "font": plot_params.axes_font},
+            legend_title={"text": legend_title, "font": plot_params.legend_font},
+        )
+
+
+PLOT_RETURN_TYPE = dp.Plot | dp.Select | dp.Group | dp.Toggle
+
+
+class SinglePlotGenerator(ABC):
+    """Generate a plot using a single BacktestResults."""
+
+    @staticmethod
+    @abstractmethod
+    def create(
+        backtest_results: BacktestResults, plot_params: PlotParameters
+    ) -> PLOT_RETURN_TYPE:
+        """Create plot object."""
+
+
+class DistributionalPlotGenerator(ABC):
+    """Generate a plot using a multiple BacktestResults."""
+
+    @staticmethod
+    @abstractmethod
+    def create(
+        backtest_results_list: Sequence[BacktestResults], plot_params: PlotParameters
+    ) -> PLOT_RETURN_TYPE:
+        """Create distributional plot object."""
+
+
+def _build_percentiles(
+    name: str,
+    df: pd.DataFrame,
+    plot_params: PlotParameters,
+) -> go.Figure:
+    mean = df.mean(axis=1)
+    mid = df.quantile(0.5, axis=1)
+    upper = df.quantile(0.75, axis=1)
+    lower = df.quantile(0.25, axis=1)
+    traces = [
+        go.Scatter(
+            name="Mean " + name,
+            x=df.index,
+            y=mean,
+            mode="lines",
+            line=dict(color=plot_params.line_colors[0]),
+        ),
+        go.Scatter(
+            name="Median " + name,
+            x=df.index,
+            y=mid,
+            mode="lines",
+            line=dict(color=plot_params.line_colors[1]),
+        ),
+        go.Scatter(
+            name="75th Percentile",
+            x=df.index,
+            y=upper,
+            mode="lines",
+            marker=dict(color=plot_params.line_colors[2]),
+            line=dict(width=0),
+            showlegend=False,
+        ),
+        go.Scatter(
+            name="25th Percentile",
+            x=df.index,
+            y=lower,
+            marker=dict(color=plot_params.line_colors[2]),
+            line=dict(width=0),
+            mode="lines",
+            fillcolor="rgba(68, 68, 68, 0.3)",
+            fill="tonexty",
+            showlegend=False,
+        ),
+    ]
+    fig = go.Figure(traces)
+    PlotParameters.update_plotly_fig(
+        plot_params,
+        fig,
+        "Percentiles",
+        "Time",
+        name,
+        "Legend",
+    )
+    return fig
+
+
+def _build_ridgeline(
+    name: str,
+    df: pd.DataFrame,
+    plot_params: PlotParameters,
+) -> go.Figure:
+    colors = n_colors(
+        plot_params.line_colors[0],
+        plot_params.line_colors[1],
+        len(df),
+        colortype="rgb",
+    )
+    df = df.copy()
+    df["colors"] = colors
+    index = df.index
+    df = df.reset_index(drop=True)
+
+    def _make_ridgeline(series: pd.Series) -> go.Violin:
+        color = series.iloc[-1]
```

Comment:
Not sure I understand where this color is coming from? the last row in the dataframe that is added above to get a nice color gradient?

Reply:
We're iterating row by row with the `apply` call, so adding the new `color` column allows the function to get the color.